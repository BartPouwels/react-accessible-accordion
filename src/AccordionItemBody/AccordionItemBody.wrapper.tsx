import * as React from 'react';

import {
    CONTEXT_KEY as ACCORDION_CONTEXT_KEY,
    getAccordionStore,
} from '../AccordionContainer/AccordionContainer';
import {
    CONTEXT_KEY as ITEM_CONTEXT_KEY,
    getItemStore,
} from '../ItemContainer/ItemContainer';
import AccordionItemBody from './AccordionItemBody';

type AccordionItemBodyWrapperProps = React.HTMLAttributes<HTMLDivElement> & {
    hideBodyClassName: string;
};

export default class AccordionItemBodyWrapper extends React.Component<
    AccordionItemBodyWrapperProps
> {
    static contextTypes = {
        // Empty anonymous callback is a hacky 'wildcard' workaround for bypassing prop-types.
        [ACCORDION_CONTEXT_KEY]: () => null,
        [ITEM_CONTEXT_KEY]: () => null,
    };

    static defaultProps = {
        className: 'accordion__body',
        hideBodyClassName: 'accordion__body--hidden',
    };

    render() {
        const itemStore = getItemStore(this.context);
        const accordionStore = getAccordionStore(this.context);
        const { uuid } = itemStore;
        const { items, accordion } = accordionStore;
        const item = items.filter(stateItem => stateItem.uuid === uuid)[0];
        return item ? (
            <AccordionItemBody
                {...this.props}
                {...item}
                accordion={accordion}
            />
        ) : null;
    }
}
