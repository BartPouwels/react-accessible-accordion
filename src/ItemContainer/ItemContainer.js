// @flow
import { Container } from 'unstated';
import consecutive from 'consecutive';

export type StoreState = {
    uuid: string | number,
};

let nextUuid = consecutive();
export function resetNextUuid() {
    nextUuid = consecutive();
}

class ItemContainer extends Container<StoreState> {
    state = {
        uuid: nextUuid(),
    };
}

export default ItemContainer;
