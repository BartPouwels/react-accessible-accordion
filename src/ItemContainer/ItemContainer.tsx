// tslint:disable:max-classes-per-file

import * as React from 'react';

export type UUID = string | number;

export type ProviderProps = {
    children?: React.ReactNode;
    uuid: UUID;
};

export type ItemContainer = {
    uuid: UUID;
};

// Arbitrary, but ought to be unique to avoid context namespace clashes.
export const CONTEXT_KEY = 'react-accessible-accordion@ItemContainer';

export class Provider extends React.Component<ProviderProps> {
    static childContextTypes = {
        // Empty anonymous callback is a hacky 'wildcard' workaround for bypassing prop-types.
        [CONTEXT_KEY]: () => null,
    };

    getChildContext() {
        const { uuid } = this.props;
        const context: ItemContainer = {
            uuid,
        };

        return {
            [CONTEXT_KEY]: context,
        };
    }

    render() {
        return this.props.children || null;
    }
}

type ConsumerProps = {
    children(container: ItemContainer): React.ReactNode;
};

export class Consumer extends React.Component<ConsumerProps> {
    static contextTypes = {
        // Empty anonymous callback is a hacky 'wildcard' workaround for bypassing prop-types.
        [CONTEXT_KEY]: () => null,
    };

    render() {
        return this.props.children(this.context[CONTEXT_KEY]);
    }
}

export const getItemStore = (context): ItemContainer | undefined => {
    return context[CONTEXT_KEY];
};
