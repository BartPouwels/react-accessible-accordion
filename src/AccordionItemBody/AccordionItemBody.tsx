import { default as classnames } from 'classnames';
import * as React from 'react';
import { UUID } from '../ItemContainer/ItemContainer';

type AccordionItemBodyProps = React.HTMLAttributes<HTMLDivElement> & {
    hideBodyClassName: string;
    uuid: UUID;
    expanded: boolean;
    disabled: boolean;
    accordion: boolean;
};

const AccordionItemBody = (props: AccordionItemBodyProps): JSX.Element => {
    const {
        className,
        hideBodyClassName,
        uuid,
        expanded,
        disabled,
        accordion,
        ...rest
    } = props;

    const role = accordion ? 'tabpanel' : null;

    return (
        <div
            id={`accordion__body-${uuid}`}
            className={classnames(className, {
                [hideBodyClassName]: !expanded,
            })}
            aria-hidden={!expanded}
            aria-labelledby={`accordion__title-${uuid}`}
            role={role}
            {...rest}
        />
    );
};

export default AccordionItemBody;
