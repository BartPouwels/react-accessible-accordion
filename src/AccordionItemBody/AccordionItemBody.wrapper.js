// @flow

import React, { Component } from 'react';
import type { ElementProps } from 'react';

import { Subscribe } from 'unstated';
import {
    Consumer,
    type AccordionContainer,
} from '../AccordionContainer/AccordionContainer';
import ItemContainer from '../ItemContainer/ItemContainer';
import AccordionItemBody from './AccordionItemBody';

type AccordionItemBodyWrapperProps = ElementProps<'div'> & {
    hideBodyClassName: string,
};

class AccordionItemBodyWrapper extends Component<
    AccordionItemBodyWrapperProps,
> {
    static defaultProps = {
        className: 'accordion__body',
        hideBodyClassName: 'accordion__body--hidden',
    };

    renderItemBody = (itemStore: ItemContainer) => {
        const { uuid } = itemStore.state;
        const { items, accordion } = this.accordionStore;
        const item = items.filter(stateItem => stateItem.uuid === uuid)[0];
        return item ? (
            <AccordionItemBody
                {...this.props}
                {...item}
                accordion={accordion}
            />
        ) : null;
    };

    accordionStore: AccordionContainer;

    renderAccordionChildren = (accordionStore: AccordionContainer) => {
        this.accordionStore = accordionStore;
        return (
            <Subscribe to={[ItemContainer]}>{this.renderItemBody}</Subscribe>
        );
    };

    render() {
        return <Consumer>{this.renderAccordionChildren}</Consumer>;
    }
}

export default AccordionItemBodyWrapper;
