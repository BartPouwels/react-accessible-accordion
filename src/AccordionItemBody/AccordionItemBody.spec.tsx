import { mount } from 'enzyme';
import * as React from 'react';
import {
    Item,
    Provider as AccordionProvider,
} from '../AccordionContainer/AccordionContainer';
import { Provider as ItemProvider } from '../ItemContainer/ItemContainer';
import AccordionItemBody from './AccordionItemBody.wrapper';

describe('AccordionItemBody', () => {
    function mountItem(Node) {
        const item: Item = {
            uuid: 0,
            expanded: false,
            disabled: false,
        };
        return mount(
            <AccordionProvider accordion={true} items={[item]}>
                <ItemProvider uuid={item.uuid}>{Node}</ItemProvider>
            </AccordionProvider>,
        );
    }

    it('renders correctly with min params', () => {
        const wrapper = mountItem(
            <AccordionItemBody>
                <div>Fake body</div>
            </AccordionItemBody>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with different className', () => {
        const className = 'className';
        const wrapper = mountItem(<AccordionItemBody className={className} />);
        expect(wrapper.find('div').hasClass(className)).toEqual(true);
    });

    it('renders correctly with different hideBodyClassName', () => {
        const hideBodyClassName = 'hideBodyClassName';
        const wrapper = mountItem(
            <AccordionItemBody hideBodyClassName={hideBodyClassName} />,
        );
        expect(wrapper.find('div').hasClass(hideBodyClassName)).toEqual(true);
    });

    it('respects arbitrary user-defined props', () => {
        const wrapper = mountItem(<AccordionItemBody lang="en" />);
        expect(wrapper.find('div').instance().lang).toEqual('en');
    });

    it('does not render if no associated item found in context', () => {
        const wrapper = mount(
            <AccordionProvider accordion={true} items={[]}>
                <ItemProvider uuid="foo">
                    <AccordionItemBody>
                        <div data-enzyme={true}>Hello World</div>
                    </AccordionItemBody>
                </ItemProvider>
            </AccordionProvider>,
        );

        expect(wrapper.find('div[data-enzyme]').length).toEqual(0);
    });
});
