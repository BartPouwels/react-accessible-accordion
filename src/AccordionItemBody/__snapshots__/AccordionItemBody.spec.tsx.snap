// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItemBody renders correctly with min params 1`] = `
<div
  aria-hidden={true}
  aria-labelledby="accordion__title-0"
  className="accordion__body accordion__body--hidden"
  id="accordion__body-0"
  role="tabpanel"
>
  <div>
    Fake body
  </div>
</div>
`;
